Add new trigger user interaction

JaJuMa opened the pull request:
### Description:

This PR is for adding a new trigger "User Interaction" in Matomo Tag Manager.

Idea with this trigger is to defer loading and processing tags until the user interacts with the page.
Specifically trigger the tags only when of these events is fired:  `"touchstart","mouseover","wheel","scroll","keydown"`

Deferring a tag via this trigger will prevent 

- page rendering being blocked by the tag and
- negative effects due to the tag on Core Web Vitals.

The idea is inspired by [Hyvä Themes](https://hyva.io) [[Source](https://docs.hyva.io/hyva-themes/writing-code/patterns/loading-external-javascript.html)]

![image](https://user-images.githubusercontent.com/12202801/207511428-3dcea806-0dd7-45af-bcd4-ab00985c0a13.png)


### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

Commit:
Add new trigger user interaction

JaJuMa committed:
Merge branch '4.x-dev' into new-trigger-user-interaction

JaJuMa commented:
@matomo-org/core-reviewers Anyone has a hint on what need to be done to get the Travis CI checks pass? :-)

AltamashShaikh commented:
@JaJuMa You need to add your new tag entry in `TagManager/tests/System/expected/test_webContext__TagManager.getAvailableTriggerTypesInContext.xml` 

To do see this doc [link](https://developer.matomo.org/guides/tests-php) 
You can run this command from your Matomo directory ` ./console tests:run plugins/TagManager/tests/System/APITest.php` which will add the new file in `TagManager/tests/System/processed/test_webContext__TagManager.getAvailableTriggerTypesInContext.xml` which you can copy to `TagManager/tests/System/expected/test_webContext__TagManager.getAvailableTriggerTypesInContext.xml` and System test cases should pass.

Alternatively you can sync the file from Travis using this command as per this [doc](https://developer.matomo.org/guides/tests-system#downloading-processed-files-from-travis-to-fix-system-tests)
`./console development:sync-system-test-processed {BUILD_NUMBER} --expected`

Note the build number changes after every push, like for the recent cimmit the build-number its `2198`

![Screenshot from 2022-12-16 12-15-01](https://user-images.githubusercontent.com/14902277/208038973-f5c677d7-d649-443c-ad8c-ee2e06bc3d65.png)

AltamashShaikh commented:
@JaJuMa Similarly when you are done fixing the System test, you can fix the UI tests using this doc [link](https://developer.matomo.org/guides/tests-ui#sync-command)

Commit:
Fix trivas CL

JaJuMa commented:
@AltamashShaikh Thanks a lot for your quick reply and support!
Tried to follow your steps and fixed 3 of the 5 Travis Tests:

![image](https://user-images.githubusercontent.com/12202801/208090271-4e8b6268-5558-4df4-b7fd-1ee7d0a37052.png)

But when trying to follow your second comment for fixing the UI tests we are getting this error:

![image](https://user-images.githubusercontent.com/12202801/208090099-885b30e2-fa84-4872-a7a1-1886b24e257e.png)

What are we doing wrong here?

UPDATE: Maybe found the problem, needed to define the correct repo in `./console tests:sync-ui-screenshots {buildnumber}`
 command. Updated PR to re-run CI test

Commit:
Fixed UI test failure

AltamashShaikh commented:
> ./console tests:sync-ui-screenshots

@JaJuMa The correct command would be `./console tests:sync-ui-screenshots -r matomo-org/tag-manager {buildNumber`}

JaJuMa commented:
> > ./console tests:sync-ui-screenshots
> 
> @JaJuMa The correct command would be `./console tests:sync-ui-screenshots -r matomo-org/tag-manager {buildNumber`}

@AltamashShaikh Yes, when we used this command it worked.  

But after updating the PR with the screenshots, 1 test is still failing with lots of  
 `Failed to load resource: the server responded with a status of 500 (Internal Server Error)` / `400 (Bad Request)`  
errors

AltamashShaikh commented:
@JaJuMa Is your PR ready? we can look into why the test cases are failing later and it looks unrelated to this change

JaJuMa commented:
@AltamashShaikh Thanks for checking the failing tests and yes, PR is ready from our side.

In case there is anything we should change or update, please let us know :-)

AltamashShaikh commented on Template/Trigger/UserInteractionTrigger.web.js:
```diff
@@ -0,0 +1,30 @@
+    return function (parameters, TagManager) {
+        this.setUp = function (triggerEvent) {
+
+            const eventNames = [
```
@JaJuMa Nothing wrong here, but as far as I remember this will create few problem in minification as per our code
Can you change it to `var`?

AltamashShaikh commented on Template/Trigger/UserInteractionTrigger.web.js:
```diff
@@ -0,0 +1,30 @@
+            
+            const removeListeners = () => {
+                for (var i = 0, iLen = eventNames.length; i < iLen; i++) {
+                    window.removeEventListener(eventNames[i], init);
```
We do not use `window` directly and use `windowAlias` instead, which you can define at the start using below code
```
var windowAlias = parameters.window;
```

AltamashShaikh commented on lang/en.json:
```diff
@@ -887,6 +887,8 @@
-        "UrlParameterVariableNameDescription": "For example when your page has URL parameters such as \"?lang=NZ\" and you want to get the value \"NZ\", then you need to enter \"lang\"."
+        "UrlParameterVariableNameDescription": "For example when your page has URL parameters such as \"?lang=NZ\" and you want to get the value \"NZ\", then you need to enter \"lang\".",
+        "UserInteractionTriggerName": "User Interaction",
+        "UserInteractionTriggerDescription": "Triggered when user interacts with the site."
```
@JaJuMa Good to mention what all interactions can trigger this

AltamashShaikh replied:
Eg: Triggered when any of the following event touch,move,scroll or mouse click is executed by the user on site.

AltamashShaikh commented on Template/Trigger/UserInteractionTrigger.web.js:
```diff
@@ -0,0 +1,30 @@
+                "keydown"
+            ];
+            
+            const init = () => {
```
Similarly make changes here and below
```
var init = function() {
};
```

Commit:
Fix standard code matomo

Commit:
Fix standard code matomo

AltamashShaikh commented on Template/Trigger/UserInteractionTrigger.web.js:
```diff
@@ -0,0 +1,32 @@
+
+            var windowAlias = parameters.window;
+            
+            var init = () => {
```
can we update this to
```
var init = function () {
        triggerEvent({event: 'UserInteraction'});
        removeListeners();
      };
```

AltamashShaikh replied:
@JaJuMa This is pending

JaJuMa replied:
@AltamashShaikh Sorry & updated

AltamashShaikh replied:
Thanks :+1:

Commit:
Fix typo and update UI screenshot

JaJuMa commented:
@AltamashShaikh Thanks for your review and comments! 👍 

PR is updated follow your ideas.

Commit:
Fix standard code matomo

AltamashShaikh commented on Template/Trigger/UserInteractionTrigger.web.js:
```diff
@@ -0,0 +1,32 @@
+                removeListeners();
+            };
+            
+            var removeListeners = () => {
```
@JaJuMa You missed the change here

AltamashShaikh reviewed (commented):
@JaJuMa Apart from 1 minor comment the PR looks okay to me, Good work :+1: 

@snake14 Can you also review this once?

Commit:
Fix standard code matomo

AltamashShaikh commented on lang/en.json:
```diff
@@ -887,6 +887,9 @@
+        "UrlParameterVariableNameDescription": "For example when your page has URL parameters such as \"?lang=NZ\" and you want to get the value \"NZ\", then you need to enter \"lang\".",
+        "UserInteractionTriggerName": "User Interaction",
+        "UserInteractionTriggerDescription": "Triggered when user interacts with the site, i. e. touch, move, scroll or mouse click event is executed by the user on site.",
+        "UserInteractionTriggerHelp": "Triggered when any of the following event touch,move,scroll or mouse click is executed by the user on site."
```
@JaJuMa Thanks for all the changes the PR looks good now
```suggestion
        "UserInteractionTriggerHelp": "Triggered when any of the following event touch, move, scroll or mouse click is executed by the user on site."
```

snake14 replied:
Altamash's suggested change of adding spaces should probably be applied as well as updating [line 420 of tests/System/expected/test_webContext__TagManager.getAvailableTriggerTypesInContext.xml](https://github.com/matomo-org/tag-manager/blob/b1bad3fd93ccfddc2312651ca5f01029e8028132/tests/System/expected/test_webContext__TagManager.getAvailableTriggerTypesInContext.xml#L420) to match.

AltamashShaikh reviewed (approved):
left 1 suggestion, rest all good

sgiehl commented:
@AltamashShaikh it looks like the screenshots that were updated in this PR were added to git directly and not using git lfs

AltamashShaikh commented:
> @AltamashShaikh it looks like the screenshots that were updated in this PR were added to git directly and not using git lfs

Ahh Thanks @sgiehl I will get this fixed

AltamashShaikh commented:
@JaJuMa Can you do this steps
`git lfs install`
`git lfs migrate import --no-rewrite tests/UI/expected-screenshots/ContainerTrigger_create_new.png`
`git lfs migrate import --no-rewrite tests/UI/expected-screenshots/ContainerTrigger_trigger_none_exist_yet_create_now.png`
`git lfs migrate import --no-rewrite tests/UI/expected-screenshots/ContainerTag_create_advanced_firetrigger_popup.png`

and later git add and commit ?
If you still face issues let me know

snake14 commented on Template/Trigger/UserInteractionTrigger.php:
```diff
@@ -0,0 +1,44 @@
+
+    public function getHelp()
+    {
+        return parent::getHelp();
```
This may not be necessary. It looks like the edit form is using both the description and help text. Unless the help text is changed to be less like the description, it just looks like it's saying the same thing twice. Any thoughts @AltamashShaikh ?
![image](https://user-images.githubusercontent.com/11394099/209032010-93311ebe-fd1f-4958-a29c-b4d12c2e6f3b.png)

snake14 replied:
One other thought is that the help text should mention that this trigger only triggers once. Until I tested it and looked more closely at the code, I thought that it would trigger every time one of those events occurred instead of just the first time for any of them.

AltamashShaikh replied:
Yes can be removed

snake14 replied:
Thank you @JaJuMa . I think that new help text looks good and thank you for making the expected XML match. I think that this PR will be ready to merge once the [screenshots are added to git lfs](https://github.com/matomo-org/tag-manager/pull/582#issuecomment-1361389136).

JaJuMa replied:
@AltamashShaikh @sgiehl @snake14 Screenshots are added to git lfs now.
Thanks again for all your great and speedy support getting this done!

snake14 reviewed (changes requested):
Looks pretty good. Aside from adding the screenshots to git lfs, I only left a couple minor comments that might need to be addressed.

Commit:
Update help text

Commit:
tests/UI/expected-screenshots/ContainerTrigger_create_new.png: convert to Git LFS

Commit:
tests/UI/expected-screenshots/ContainerTrigger_trigger_none_exist_yet_create_now.png: convert to Git LFS

Commit:
tests/UI/expected-screenshots/ContainerTag_create_advanced_firetrigger_popup.png: convert to Git LFS

snake14 reviewed (approved):
Looks good. Thank you for all of your effort :+1:

snake14 commented:
@AltamashShaikh Are we good to merge this PR?

Yes we can
